lib/srv/db/server.go

```diff
@@ -689,7 +699,36 @@ func (s *Server) handleConnection(ctx context.Context, conn net.Conn) error {
 		return trace.Wrap(err)
 	}
 
-	err = engine.HandleConnection(ctx, sessionCtx, conn)
+	if err := engine.InitializeConnection(clientConn, sessionCtx); err != nil {
+		return trace.Wrap(err)
+	}
+
+	defer func() {
+		if r := recover(); r != nil {
+			s.log.Warnf("Recovered while handling DB connection from %v: %v.", clientConn.RemoteAddr(), r)
+			err = trace.BadParameter("failed to handle client connection")
+		}
+		if err != nil {
+			engine.SendError(err)
+		}
+	}()
```

Comment:
Should this be moved above right after the `engine` is created so if an error happens during monitorConn or InitializeConnection, it would also be sent to the client properly?

Reply:
As I mentioned below, an error cannot be sent before `InitializeConnection()` is called.